File: lib/domain/cultivation_plan/gateways/plan_save_user_fertilize_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      class PlanSaveUserFertilizeGateway
        # @return [Dtos::PlanSaveUserFertilizeSnapshot, nil]
        def find_by_user_id_and_source_fertilize_id(user_id:, source_fertilize_id:)
          raise NotImplementedError
        end

        # @param user_id [Integer]
        # @param attributes [Hash]
        # @return [Dtos::PlanSaveUserFertilizeSnapshot]
        # @raise [Domain::Shared::Exceptions::RecordInvalid]
        def create(user_id:, attributes:)
          raise NotImplementedError
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/plan_save_user_fertilize_gateway.rs
//! Ruby: `Domain::CultivationPlan::Gateways::PlanSaveUserFertilizeGateway`

use crate::cultivation_plan::dtos::PlanSaveUserFertilizeSnapshot;
use crate::shared::attr::AttrMap;

pub trait PlanSaveUserFertilizeGateway: Send + Sync {
    fn find_by_user_id_and_source_fertilize_id(
        &self,
        user_id: i64,
        source_fertilize_id: i64,
    ) -> Result<Option<PlanSaveUserFertilizeSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn create(
        &self,
        user_id: i64,
        attributes: AttrMap,
    ) -> Result<PlanSaveUserFertilizeSnapshot, Box<dyn std::error::Error + Send + Sync>>;
}
